### PR TITLE
COMPAS Plugins

### PR DESCRIPTION
**Pull request by gonzalocasas**
This PR adds a plugin system to COMPAS according to the specs in https://github.com/compas-dev/compas/issues/576
As an example, this also adds a first set of `pluggable` interfaces for boolean operations, and their corresponding plugin implementations for `compas_rhino` using Rhino SDK. Additionally, I am sending a PR to `compas_cgal` repo to add plugin support to the boolean operators there.

I didn't find a nice and simple way to add unit tests yet thou.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

Since I added quite a bit of docs, and it's annoying to review RST, here's a screenshot of the built dev guide with the plugin content:
docs

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,8 +9,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
-* Added `compas.datastructures.Mesh.transform_numpy`.
-* Added exceptions module.
+* Added plugin system based on decorators: `compas.plugins.pluggable` & `compas.plugins.plugin`.
+* Added `compas_rhino` implementation of the boolean operation pluggable interfaces (union/difference/intersection).
+* Added `compas.datastructures.Mesh.transform_numpy`
```
Comment (beverlylytle) [inline CHANGELOG.md]: Full stop?

File: src/compas/plugins.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        return self._cache[module_name]
+
+
+def select_plugin(extension_point_url, manager):
+    if manager.DEBUG:
+        print('Extension Point URL {} invoked. Will select a matching plugin'.format(extension_point_url))
+
+    plugins = manager.registry.get(extension_point_url)
+    if plugins:
+        for plugin in plugins:
```
Comment (beverlylytle) [inline src/compas/plugins.py]: One level of nesting could be eliminated with:
```
for plugin in plugins or []:

**Review by beverlylytle — APPROVED** [review]
LGTM!

File: src/compas/plugins.py (hunk 2 above)
Comment (gonzalocasas) [reply src/compas/plugins.py]: oh thx! love this construct! never thought about it

Comment (gonzalocasas) [reply src/compas/plugins.py]: done on https://github.com/compas-dev/compas/pull/598/commits/688beb5ec2d7a0b34111106fe5a04e0703f06a1e

**Pull request merged**